core/controllers/async_download.py

```diff
@@ -0,0 +1,147 @@
+from core.aws import _S3_CLIENT
+from core.controllers.download import download
+
+
+def async_download(
+    email_address: str,
+    file_format: str,
+    funds: list[str] | None = None,
+    organisations: list[str] | None = None,
+    regions: list[str] | None = None,
+    rp_start: str | None = None,
+    rp_end: str | None = None,
+    outcome_categories: list[str] | None = None,
+):
+    """Download data, store file in S3 and send an email to the user with the download link."""
+
+    response = download(
+        file_format=file_format,
+        funds=funds,
+        organisations=organisations,
+        regions=regions,
+        rp_start=rp_start,
+        rp_end=rp_end,
+        outcome_categories=outcome_categories,
+    )
+
+    if response.status_code != 200:
+        current_app.logger.error(
+            "Error downloading data from data-store/download endpoint. Status code: {response_status}",
+            extra=dict(response_status=response.status_code),
+        )
+        return
+
+    file_format_str: str = get_file_format_from_extension(file_format)
+    file_size_str: str = get_human_readable_file_size(len(response.data))
+
+    # Upload the file to S3 and get presigned URL
+    file_obj = io.BytesIO(response.data)
+    bucket = Config.AWS_S3_BUCKET_FIND_DATA_FILES
+    find_service_base_url = Config.FIND_SERVICE_BASE_URL
+    email_template_id = Config.NOTIFY_FIND_REPORT_DOWNLOAD_TEMPLATE_ID
+
+    if not bucket:
+        current_app.logger.error("AWS_S3_BUCKET_FIND_DATA_FILES is not set.")
+    elif not find_service_base_url:
+        current_app.logger.error("FIND_SERVICE_BASE_URL is not set.")
+    elif not email_template_id:
+        current_app.logger.error("NOTIFY_TEMPLATE_ID is not set.")
+    else:
+        current_datetime = datetime.now().strftime("%Y-%m-%d-%H:%M:%S")
+        file_name = f"fund-monitoring-data-{current_datetime}.{file_format}"
+        url = upload_file(file=file_obj, bucket=bucket, object_name=file_name, expiry=86400)
+
+        # email the URL to the user
+        send_email(
+            template_id=email_template_id,
+            email_address=email_address,
+            download_url=url,
+            find_service_url=find_service_base_url + "/download",
+            file_format=file_format_str,
+            file_size_str=file_size_str,
+        )
+
+
+def upload_file(file: IO | FileStorage, bucket: str, object_name: str, expiry: int) -> str:
+    """Upload a file to an S3 bucket, generate presigned url and return it
+    :param file: File to upload,
+    :param bucket: Bucket to upload to,
+    :param object_name: S3 object name (filename),
+    :param expiry: URL expiry time in seconds,
+
+    :return: Presigned URL
+    """
+
+    _S3_CLIENT.put_object(Bucket=bucket, Key=object_name, Body=file)
+
+    url = _S3_CLIENT.generate_presigned_url(
+        "get_object",
+        Params={"Bucket": bucket, "Key": object_name},
+        ExpiresIn=expiry,
+    )
+
+    return url
```

Comment:
This function probably lives better in `core/aws`. Note that there is already an `upload_file` function there - maybe we should reuse that?

This function feels a bit overloaded - it uploads the file _and_ generates a presigned URL. These feel like different things.

I think we can either split them out (just make these two calls in `async_download`, or just rename the function to be a bit clearer? Especially because the current signature is called `upload_file` with an `expiry` argument. That sounds like the file will expire after that long - it doesn't say that's just how long the link will (should) last.

Reply:
Function split into separate functions for file upload and generating the url.